### PR TITLE
Indentation Improvements

### PR DESCRIPTION
**Pull request by martinblake**
# Main change
Indentation improvements, mainly based on levels of brackets, plus "continuation indentation" where certain symbols begin/end lines.  The results of these changes are best understood by looking at the changes/additions to `test/sample.kt`.  I've extended `test/kotlin-mode-test.el` with a test that verifies the accuracy of this file by indenting each line and checking that nothing changes.

# Extra
- Removed override of `<tab>` key mapping, which was preventing region indentation working for me.
- Added highlighting of `get`, `set`, `field` and `it` keywords.  (I realise some of this had been deliberately commented out, so happy to revert if it's still work-in-progress)

### CODE REVIEW

File: test/sample.kt [hunk 1]
```diff
@@ -287,7 +302,9 @@ class Derived() : Base() {
     override fun v() {}
 }
 
-open class AnotherDerived() : Base() {
+open class AnotherDerived()
+    : Base()
```
Comment (gregghz) [inline test/sample.kt:306]: i wonder if changes like this should be configurable. I'm not very in touch with what "standard" kotlin looks like these days so I'm not sure what is "correct" here. My gut says not this, but I suspect at the very least opinions will differ. Same thoughts on line 219.

**Review by gregghz — COMMENTED** [review]
This is a great refactor. It fixes the long standing issues with getters and setters. Like I mentioned in another comment, I'm not very in touch with what the kotlin community considers best practice for formatting. However, this PR does bring up the idea of whether or not kotlin-mode should have a strong opinion on how to format code vs. having as many styles configurable as possible.

Let me know your thoughts.

File: kotlin-mode.el [hunk 2]
```diff
@@ -324,70 +325,175 @@
...
+   determines the desired indentation level for the current line.")
+
+(cl-defmethod count-to-line-start ((counter kotlin-mode--bracket-counter))
+  "Count the brackets on the current line, starting from the cursor
+   position, and working backward, incrementing the count
+   +1 for open-brackets, -1 for close-brackets.
+   Return as soon as the overall count exceeds zero."
+  (save-excursion
+    (while (and (<= (oref counter count) 0) (not (bolp)))
+      (backward-char)
```
Comment (taku0) [inline kotlin-mode.el:394]: This should skip comment and string.

```kotlin
val x = "("
          foo()
```

**Comment by taku0** [comment]
I propose adding magic comments to where the current indentation results are not ideal, and ignoreing those lines while testing:

```kotlin
fun max(a: Int, b: Int): Int {
    if (a > b)
        return a // KNOWN_BUG
    else
        return b // KNOWN_BUG
}
```

[`swift-mode`](https://github.com/swift-emacs/swift-mode), which I maintain, takes similar strategy ([an example of `known-bug` comment](https://github.com/swift-emacs/swift-mode/blob/be8d7700cdbf47576d7c4e0a7e0855cce0fe9ad8/test/swift-files/indent/statements.swift#L849)).  Indentation mismatches on those lines are reported as warnings rather than errors, and when indentation logic is fixed, it suggests removing the comments.

**Comment by taku0** [comment]
Indentation of a multiline comment on the line 1 is incorrect.

```kotlin
/*
*/
```

Expected:

```kotlin
/*
 */
```

**Comment by taku0** [comment]
> whether or not kotlin-mode should have a strong opinion on how to format code vs. having as many styles configurable as possible.

It is a difficult problem. [`cc-mode` has a very flexible indentation engine](https://www.gnu.org/software/emacs/manual/html_node/ccmode/Indentation-Engine-Basics.html) but building it would be hard.  Many other modes, at least my indentation engine in `swift-mode`, tries to indent reasonably for many indentation styles with small set of customization variables. For example, `swift-mode` supports those styles without customization variables:

```swift
class Foo:
  Bar,
  Baz {
}

class Foo
  : Bar,
    Baz {
}

class Foo
  : Bar
  , Baz {
}
```

I suggest merging this PR for now and improving it progressively, because implementing a good indentation logic takes a long time and this indentation logic is reasonably good compared to the current one.

**Comment by martinblake** [comment]
Thanks @gregghz and @taku0 for your helpful reviews.  I'll try and address them properly when I have time in the next week or two, though I agree that it would be good to merge soon and make incremental improvements in the future.  The `KNOWN_BUG` tag seems like a good way to manage this!

**Comment by gregghz** [comment]
Thanks for the contribution @martinblake ! I'm happy to help out and review any follow up work you plan to do.

**Pull request merged**